core/src/main/scala/chisel3/Data.scala

```diff
@@ -435,27 +435,78 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
+        case (Some(a), Some(b)) => s"$a($b)"
+        case (None, None) => ""
+        case (Some(a), _) => s"$a"
+        case (_, Some(b)) => s"$b"
+      }
+    } else {
+      val (binding: Option[String], enclosure: Option[BaseModule]) = bindingToString
+        (fmtParentStr,
+          _computeName(None, None),
+          fmtSubcomponentStr(enclosure),
+          binding,
+          fmtChiselTypeStr(chiselType),
+          fmtValuesStr(vals)) match {
+        // TODO figure out which ones are redundant/impossible
+        //    parent   name     subcomp  binding  type     values
+        case (Some(a), Some(b), Some(c), Some(d), Some(e), Some(f)) => s"$a.$b.$c: $d[$e]($f)"
+        case (Some(a), Some(b), Some(c), Some(d), Some(e), None) => s"$a.$b.$c: $d[$e]"
+        case (Some(a), None, None, Some(d), Some(e), None) => s"$a: $d[$e]"
+        case (Some(a), Some(b), _, Some(d), Some(e), _) => s"$a.$b: $d[$e]"
+        case (Some(a), None, Some(c), Some(d), Some(e), None) => s"$a.$c: $d[$e]"
+        case (Some(a), Some(b), None, None, Some(e), None) => s"$a.$b: $e"
+        case (None, None, Some(c), Some(d), Some(e), Some(f)) => s"$d[$e]($f)"
+        case (_, None, None, None, Some(e), None) => s"$e"
+
+        case (Some(a), None, None, Some(d), Some(e), Some(f)) => s"$a: $d[$e]($f)"
+        case (None, None, None, None, None, None) => s""
+        // case (_, _, _, _, _, _) => s"" // TODO this should be deleted once all cases are accounted for
+      }
+    }
+  }
+
   // User-friendly representation of the binding as a helper function for toString.
   // Provides a unhelpful fallback for literals, which should have custom rendering per
   // Data-subtype.
   // TODO Is this okay for sample_element? It *shouldn't* be visible to users
-  protected def bindingToString: String = Try(topBindingOpt match {
-    case None => ""
-    case Some(OpBinding(enclosure, _)) => s"(OpResult in ${enclosure.desiredName})"
-    case Some(MemoryPortBinding(enclosure, _)) => s"(MemPort in ${enclosure.desiredName})"
-    case Some(PortBinding(enclosure)) if !enclosure.isClosed => s"(IO in unelaborated ${enclosure.desiredName})"
-    case Some(PortBinding(enclosure)) if enclosure.isClosed =>
-      DataMirror.fullModulePorts(enclosure).find(_._2 eq this) match {
-        case Some((name, _)) => s"(IO $name in ${enclosure.desiredName})"
-        case None => s"(IO (unknown) in ${enclosure.desiredName})"
-      }
-    case Some(RegBinding(enclosure, _)) => s"(Reg in ${enclosure.desiredName})"
-    case Some(WireBinding(enclosure, _)) => s"(Wire in ${enclosure.desiredName})"
-    case Some(DontCareBinding()) => s"(DontCare)"
-    case Some(ElementLitBinding(litArg)) => s"(unhandled literal)"
-    case Some(BundleLitBinding(litMap)) => s"(unhandled bundle literal)"
-    case Some(VecLitBinding(litMap)) => s"(unhandled vec literal)"
-  }).getOrElse("")
+  protected def bindingToString: (Option[String], Option[BaseModule]) =
```

Comment:
```suggestion
  @deprecated("This was never intended to be visible to user-defined types", "Chisel 3.5.0")
  protected def bindingToString: String = this.bindingToStringAndEnclosure._1.getOrElse("")
  
  private[core] def bindingToStringAndEnclosure: (Option[String], Option[BaseModule]) =
```

All of the added methods should be package private so that users can't call them. I think it is a mistake that `bindnigToString` was `protected`. Some online docs claim that `protected` scope in Scala only extends to subclasses in the same package (which may have motivated the original code), but that is not true so I think we should deprecate here.

Also, I found it difficult to come up with a good name for `bindingToStringAndEnclosure` which may be indicative of this being not a great abstraction. That being said, I understand wanting to save multiple `.topBinding` lookups. Another option is to do something like this:

```scala
// Only make the argument have a default if this is called in several places, if not I'd just make it required
private[core] def bindingToString(topBindingOpt: Option[TopBinding] = this.topBindingOpt) = ... // old implementation

// Then tweak the use slightly:
val bindingOpt = this.topBindingOpt
val binding: Option[String] = this.bindingToString(bindingOpt)
// I'm not 100% location is equivalent to your manual extraction of enclosure, but I think it is...
val enclosure: Option[BaseModule] bindingOpt.flatMap(_.location)
```

Reply:
Looks great thanks - this should really simplify the `bindingToString` API